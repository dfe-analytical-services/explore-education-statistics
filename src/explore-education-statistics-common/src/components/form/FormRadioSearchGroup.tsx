import { FormFieldsetProps } from '@common/components/form/FormFieldset';
import FormTextSearchInput from '@common/components/form/FormTextSearchInput';
import { FormFieldset } from '@common/components/form/index';
import useMounted from '@common/hooks/useMounted';
import FormRadioGroup, {
  BaseFormRadioGroup,
  FormRadioGroupProps,
} from '@common/components/form/FormRadioGroup';
import React, { useState } from 'react';

export interface FormRadioSearchGroupProps extends FormRadioGroupProps {
  searchLabel?: string;
}

const FormRadioSearchGroup = ({
  searchLabel = 'Search',
  ...props
}: FormRadioSearchGroupProps) => {
  const { isMounted } = useMounted();

  const [searchTerm, setSearchTerm] = useState('');
  const [selectedOption, setSelectedOption] = useState('');

  const {
    id,
    hint,
    legend,
    legendHidden,
    legendSize = 'm',
    error,
    name,
    onFieldsetFocus,
    onFieldsetBlur,
    options = [],
    ...groupProps
  } = props;

  const fieldsetProps: FormFieldsetProps = {
    id,
    legend,
    legendHidden,
    legendSize,
    hint,
    error,
    onFocus: onFieldsetFocus,
    onBlur: onFieldsetBlur,
  };

  if (!isMounted) {
    return (
      <FormRadioGroup
        {...props}
        name={name}
        legend={legend}
        id={id}
        options={options}
      />
    );
  }

  let filteredOptions = options;

  if (searchTerm) {
    filteredOptions = options.filter(option => {
      return (
        option.label.toLowerCase().includes(searchTerm.toLowerCase()) ||
        selectedOption.indexOf(option.value) > -1
      );
    });
  }

  return (
    <FormFieldset {...fieldsetProps} useFormId={false}>
      {options.length > 1 && (
        <FormTextSearchInput
          id={`${id}-search`}
          className="govuk-!-margin-bottom-4"
          name={`${name}-search`}
          label={searchLabel}
          width={20}
          onChange={event => setSearchTerm(event.target.value)}
          onKeyPress={event => {
            if (event.key === 'Enter') {
              event.preventDefault();
            }
          }}
        />
      )}

      <div aria-live="assertive">
        {filteredOptions.length > 0 ? (
          <BaseFormRadioGroup
            {...groupProps}
            name={name}
            id={id}
            options={filteredOptions}
            onChange={(event, option) => {
              setSelectedOption(event.target.value);
              if (props.onChange) {
                props.onChange(event, option);
              }
            }}
          />
        ) : (
          <p>No results found</p>
        )}
      </div>
    </FormFieldset>
  );
};

export default FormRadioSearchGroup;
