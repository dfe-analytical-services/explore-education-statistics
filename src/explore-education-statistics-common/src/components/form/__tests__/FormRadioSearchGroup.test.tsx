import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import FormRadioSearchGroup from '@common/components/form//FormRadioSearchGroup';
import React from 'react';

jest.mock('lodash/debounce');

describe('FormRadioSearchGroup', () => {
  test('renders list of radio buttons in correct order', () => {
    const { container } = render(
      <FormRadioSearchGroup
        name="testRadios"
        id="test-radios"
        legend="Choose options"
        options={[
          { label: 'Test radio 1', value: '1' },
          { label: 'Test radio 2', value: '2' },
          { label: 'Test radio 3', value: '3' },
        ]}
      />,
    );

    const radios = screen.getAllByLabelText(/Test radio/) as HTMLInputElement[];

    expect(radios).toHaveLength(3);
    expect(radios[0]).toHaveAttribute('value', '1');
    expect(radios[1]).toHaveAttribute('value', '2');
    expect(radios[2]).toHaveAttribute('value', '3');

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('providing a search term renders only relevant radio buttons', async () => {
    jest.useFakeTimers();

    render(
      <FormRadioSearchGroup
        name="testRadios"
        id="test-radios"
        legend="Choose options"
        searchLabel="Search options"
        options={[
          { label: 'Test radio 1', value: '1' },
          { label: 'Test radio 2', value: '2' },
          { label: 'Test radio 3', value: '3' },
        ]}
      />,
    );

    const searchInput = screen.getByLabelText('Search options');

    userEvent.type(searchInput, '2');

    jest.runAllTimers();

    const radios = screen.getAllByLabelText(/Test radio/) as HTMLInputElement[];

    expect(radios).toHaveLength(1);
    expect(radios[0]).toHaveAttribute('value', '2');
  });

  test('does not throw error if search term that is invalid regex is used', async () => {
    jest.useFakeTimers();

    render(
      <FormRadioSearchGroup
        name="testRadios"
        id="test-radios"
        legend="Choose options"
        searchLabel="Search options"
        options={[
          { label: 'Test radio 1', value: '1' },
          { label: 'Test radio 2', value: '2' },
          { label: 'Test radio 3', value: '3' },
        ]}
      />,
    );

    const searchInput = screen.getByLabelText('Search options');

    userEvent.type(searchInput, '[');

    expect(() => {
      jest.runAllTimers();
    }).not.toThrow();

    const radios = screen.queryAllByLabelText(
      /Test radio/,
    ) as HTMLInputElement[];

    expect(radios).toHaveLength(0);
  });

  test('providing a search term does not remove radio buttons that have already been checked', async () => {
    jest.useFakeTimers();

    render(
      <FormRadioSearchGroup
        name="testRadios"
        id="test-radios"
        legend="Choose options"
        searchLabel="Search options"
        options={[
          { label: 'Test radio 1', value: '1' },
          { label: 'Test radio 2', value: '2' },
          { label: 'Test radio 3', value: '3' },
        ]}
      />,
    );

    const radio = screen.getByLabelText('Test radio 1') as HTMLInputElement;

    fireEvent.click(radio);

    const searchInput = screen.getByLabelText('Search options');

    userEvent.type(searchInput, '2');

    jest.runAllTimers();

    const radios = screen.getAllByLabelText(/Test radio/) as HTMLInputElement[];

    expect(radios).toHaveLength(2);
    expect(radios[0]).toHaveAttribute('value', '1');
    expect(radios[1]).toHaveAttribute('value', '2');
  });

  test('shows a message if no search results are found', async () => {
    jest.useFakeTimers();

    render(
      <FormRadioSearchGroup
        name="testRadios"
        id="test-radios"
        legend="Choose options"
        searchLabel="Search options"
        options={[
          { label: 'Test radio 1', value: '1' },
          { label: 'Test radio 2', value: '2' },
          { label: 'Test radio 3', value: '3' },
        ]}
      />,
    );

    const searchInput = screen.getByLabelText('Search options');

    userEvent.type(searchInput, 'Not there');

    jest.runAllTimers();

    const radios = screen.queryAllByLabelText(
      /Test radio/,
    ) as HTMLInputElement[];

    expect(radios).toHaveLength(0);
    expect(screen.getByText('No results found')).toBeInTheDocument();
  });
});
