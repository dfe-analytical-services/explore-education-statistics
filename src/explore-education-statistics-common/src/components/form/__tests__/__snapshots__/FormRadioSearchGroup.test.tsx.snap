// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRadioSearchGroup renders list of radio buttons in correct order 1`] = `
<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-radios"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Choose options
    </legend>
    <label class="govuk-label"
           for="test-radios-search"
    >
      Search
    </label>
    <input name="testRadios-search"
           class="govuk-input govuk-!-margin-bottom-4 searchInput govuk-input--width-20"
           id="test-radios-search"
           type="text"
           value
    >
    <div aria-live="assertive">
      <div class="govuk-radios">
        <div class="govuk-radios__item"
             data-testid="Radio item for Test radio 1"
        >
          <input class="govuk-radios__input"
                 id="test-radios-1"
                 name="testRadios"
                 type="radio"
                 value="1"
          >
          <label class="govuk-label govuk-radios__label"
                 for="test-radios-1"
          >
            Test radio 1
          </label>
        </div>
        <div class="govuk-radios__item"
             data-testid="Radio item for Test radio 2"
        >
          <input class="govuk-radios__input"
                 id="test-radios-2"
                 name="testRadios"
                 type="radio"
                 value="2"
          >
          <label class="govuk-label govuk-radios__label"
                 for="test-radios-2"
          >
            Test radio 2
          </label>
        </div>
        <div class="govuk-radios__item"
             data-testid="Radio item for Test radio 3"
        >
          <input class="govuk-radios__input"
                 id="test-radios-3"
                 name="testRadios"
                 type="radio"
                 value="3"
          >
          <label class="govuk-label govuk-radios__label"
                 for="test-radios-3"
          >
            Test radio 3
          </label>
        </div>
      </div>
    </div>
  </fieldset>
</div>
`;
